sherpa/models/model.py

```diff
@@ -731,6 +731,75 @@ def __setattr__(self, name: str, val) -> None:
         for alias in val.aliases:
             self._par_index[alias] = val
 
+    def get_parts(self,
+                  include_composites : bool = True,
+                  remove_duplicates : bool = False) -> Sequence[Model]:
+        """Return the parts of the model.
+
+        Parameters
+        ----------
+        include_composites : bool, optional
+            If True, include the components of any composite models.
+            If False, only include the top-level components.
+        remove_duplicates : bool, optional
+            If True, remove any duplicate components. This is
+            useful if the same component is used in multiple
+            places in the model.
+
+        Returns
+        -------
+        modellist : list
+            The parts of the model.
+        """
+        # Note:
+        # include_composites and remove_duplicates is not used in the base class,
+        # but it is used in the CompositeModel class.
+        return [self]
+
+    def get_components_by_name(self, name: str) -> list[Model]:
+        """Return the components of the model with the given name.
+
+        Parameters
+        ----------
+        name : str
+            The name of the component to return.
+
+        Returns
+        -------
+        modellist : list
+            The components of the model with the given name.
+        """
+        parts = [p for p in self.get_parts(remove_duplicates=True) if p.name == name]
+        if len(parts) == 0:
+            raise KeyError(f"No components found with name '{name}'")
+        return parts
+
+    def get_components_by_class(self, cls: type, subclass_ok: bool = True) -> list[Model]:
+        """Return the components of the model with the given type.
+
+        Parameters
+        ----------
+        type : type
+            The type of the component to return.
+        subclass_ok : bool, optional
+            If True, return components that are subclasses of the given
+            type. If False, only return components that are exactly of
+            the given type.
+
+        Returns
+        -------
+        list[Model]
+            The components of the model with the given type.
+        """
+        if subclass_ok:
+            parts = [p for p in self.get_parts(remove_duplicates=True) if isinstance(p, cls)]
+        else:
+            parts = [p for p in self.get_parts(remove_duplicates=True) if type(p) is cls]
```

Comment:
This could be something like (untested)

```python
def checkfunc(p: Model) -> bool:
    if subclass_ok:
        return isinstance(p, cls)
    return type(p) is cls

parts = [p for p in self.get_parts(remove_duplicates=True) if checkfunc(p)]
```

We could instead define `checkfunc` "within" the `subclass_ok` check, which would be more efficient but I think it would be harder to follow and we don't care too much about efficiency here.

Reply:
It could be but that's actually 5 lines instead of 4 and another level of indirection. It's not reused in other parts of the code so I don't see a benefit to abstract it out here.
However, if @DougBurke insists I'm happy to do it, I don't feel very strongly about it.

Reply:
It's not so much the characters saved, but I think it's more obvious what is different when `subclass_ok` is set or not. With the original  version I have to check the `parts = [..]` line twice to spot the difference, where in the new version it is more-obvious (to me) that `subclass_ok` only changes how the model and class are compared.